symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+TIMEOUT = 7200
 
 thisdir = os.path.dirname(__file__)
 log = logging.getLogger()
 
 
-class Win64ProcessError(Exception):
-    pass
-
-
-# Don't run continuously for more than 2 hours.
-MAX_RUNTIME_SECS = 7200
+def get_type(data):
+    # PDB v7
+    if data.startswith(b"Microsoft C/C++ MSF 7.00"):
+        return "pdb-v7"
+    # PDB v2
+    if data.startswith(b"Microsoft C/C++ program database 2.00"):
+        return "pdb-v2"
+    # DLL
+    if data.startswith(b"MZ"):
+        return "dll"
+    # CAB
+    if data.startswith(b"MSCF"):
+        return "cab"
 
+    return "unknown"
 
-def timed_out(start):
-    return (datetime.datetime.now() - start).total_seconds() > MAX_RUNTIME_SECS
 
-
-def fetch_symbol(debug_id, debug_file):
-    '''
-    Attempt to fetch a PDB file from Microsoft's symbol server.
-    '''
-    # TODO: figure out how to request compressed symbols nowadays.
-    url = urlparse.urljoin(MICROSOFT_SYMBOL_SERVER,
-                           os.path.join(debug_file,
-                                        debug_id,
-                                        debug_file))
-    try:
-        r = requests.get(url,
-                         headers={'User-Agent': USER_AGENT})
-        if r.status_code == 200:
-            return r.content
-    except requests.exceptions.RequestException as e:
-        log.warn('Error fetching symbol file "%s/%s": %s' % (debug_id, debug_file, str(e)))
-    return None
-
-
-def fetch_symbol_and_decompress(tmpdir, debug_id, debug_file):
-    '''
-    Attempt to fetch a PDB file from Microsoft's symbol server, then
-    write a decompressed version to tmpdir.
-    Returns the filename if successful, or None if unsuccessful.
-    '''
-    data = fetch_symbol(debug_id, debug_file)
-    if not data:
-        return None
-    # Check for PDB/PE signature.
-    if data.startswith(b'Microsoft C/C++ MSF 7.00') or data.startswith(b'MZ'):
-        # Uncompressed file, just write it out and return it.
-        path = os.path.join(tmpdir, debug_file.lower())
-        with open(path, 'wb') as f:
-            f.write(data)
-        return path
-    if not data.startswith(b'MSCF'):
-        # Not a cabinet file.
-        return None
-    path = os.path.join(tmpdir, debug_file[:-1] + '_')
-    with open(path, 'wb') as f:
-        f.write(data)
-    try:
-        # Decompress it
-        subprocess.check_call(['cabextract', '-L', '-d', tmpdir, path],
-                              stdout=open(os.devnull, 'wb'))
-        os.unlink(path)
-        return os.path.join(tmpdir, debug_file.lower())
-    except subprocess.CalledProcessError:
-        return None
-
-
-def fetch_and_dump_symbols(tmpdir, debug_id, debug_file,
-                           code_id=None, code_file=None):
-    pdb_path = fetch_symbol_and_decompress(tmpdir, debug_id, debug_file)
-    if pdb_path is None:
-        return None
-    bin_path = None
-    while True:
+async def server_has_file(client, server, filename):
```

Comment:
Nit: I'd introduce a class where all these functions are defined, so we can avoid passing down the same arguments in all funcs